### ISSUE TITLE
XLA commonly aborts with nvidia error

### ISSUE THREAD

**Issue by joeyearsley** [tensorflow/tensorflow#22119]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: 1.10
- **Python version**: 3.5
- **Bazel version (if compiling from source)**: 0.15
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.2 / 7.2.1.38
- **GPU model and memory**: Titan
- **Exact command to reproduce**: N/A

### Describe the problem
When running XLA via the session config we commonly hit the following error:

`2018-09-06 17:28:38.814423: F tensorflow/stream_executor/cuda/cuda_dnn.cc:211] Check failed: status == CUDNN_STATUS_SUCCESS (7 vs. 0)Failed to set cuDNN stream.
`

However, sometimes we get further than this error and see significant speed up with XLA and the loss decreasing.

What could be causing this error?

The docker image is located at: https://hub.docker.com/r/joeyearsley/tf-builds/

I can't diverge the exact script, however I do have the following:
- NCCL for multi-gpu reduction like TF-Benchmark's Variable Manager
- `tf.data` APIs including the string handles to alter between train and validation datasets.

**Comment by joeyearsley** [comment]
@jlebar I got around to trying XLA again after #16045, works nicely when I get past this error, however stumped as this errors pops up frequently. Around 80% likely when running a script compared to the 20% which actually get past this error and continue.

**Comment by jlebar** [comment]
> Check failed: status == CUDNN_STATUS_SUCCESS (7 vs. 0)Failed to set cuDNN stream.

It's possible that there's an error earlier in execution, and this is just fallout.  Once there's some error on a stream, any operation on the stream will fail.

Could you look through the logs carefully for warnings?  And/or run with CUDA_LAUNCH_BLOCKING=1, which may make it easier to find "up-stream" errors.

**Comment by joeyearsley** [comment]
Running it with that command still results in the same error.

It runs consistently when I include these flags `TF_XLA_FLAGS="--parallel_check_failfast=false"`

**Comment by joeyearsley** [comment]
I changed from using the session approach to the scoped approach to try and narrow down potential places the error could occur. 

It seems to be when I have `opt.compute_gradients(loss, params)` compiled into XLA code that I see a weird 3x speed-up but the loss doesn't decrease, and occasionally get the `CUDNN_STATUS_SUCCESS`.

If I just have the jit_scope around my model then it runs fine.

**Comment by jlebar** [comment]
> It runs consistently when I include these flags TF_XLA_FLAGS="--parallel_check_failfast=false"

The parallel_check_op is deleted at HEAD because it was unused.  If it were affecting things, then you should see in your logs `failfast on first parallel-check failure`.  If you don't see that, I don't think this flag can possibly be changing anything.

Can you confirm that you have looked carefully through the logs when running with CUDA_LAUNCH_BLOCKING and do not notice anything else suspicious?  Is it possible to post (or send privately) your full log, or if you're not comfortable with that, the log of all lines that start with `W` or `E` (i.e., all warnings and errors)?

> It seems to be when I have opt.compute_gradients(loss, params) compiled into XLA code that I see a weird 3x speed-up but the loss doesn't decrease

Weird.

It could totally be a bug in XLA (or cudnn), but it may be pretty hard to debug without the ability to reproduce.

**Comment by joeyearsley** [comment]
Just to double check, I can gather the logs by setting `tf.logging.set_verbosity('DEBUG')`? - I missed changing this previously..

I'll try making the exact copy of my script with the mnist_xla script but making use of all things like nccl for multi-gpu.

**Comment by jlebar** [comment]
>  I can gather the logs by setting tf.logging.set_verbosity('DEBUG')? - I missed changing this previously..

That would be sufficient.

> I'll try making the exact copy of my script with the mnist_xla script but making use of all things like nccl for multi-gpu.

OK, thanks!

**Comment by joeyearsley** [comment]
No errors or warnings when logging and `CUDA_LAUNCH_BLOCKING=1`.

Working on a mnist_xla script now

**Comment by joeyearsley** [comment]
https://github.com/joeyearsley/mnist-xla/tree/master

**Comment by jlebar** [comment]
@d0k spent some time looking at this.  He was able to reproduce the high loss, with both XLA:GPU *and* XLA:CPU.  Still unclear what's going on.

One question for you is, do you have a way we can reproduce the `Check failed: status == CUDNN_STATUS_SUCCESS (7 vs. 0)` error?  That may be separate.

**Comment by joeyearsley** [comment]
I’ll try reproducing that result with the mnist example.

We’ve also been seeing weird things with TF 1.10 and cuda streams but that’ll be another issue created after a bit more investigation - the CUDA error might be more related to that.

**Comment by joeyearsley** [comment]
Trying to reproduce the CUDNN error, so far no luck as it is extremely stochastic.

Seems to be happening more frequently when there's a batch size which causes:
`Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.06GiB. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.`

and when there are multiple GPUs.

Finding it hard to make a reproducible example.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by joeyearsley** [comment]
Anymore information on the XLA not working?

**Comment by jlebar** [comment]
We haven't forgotten about this, but I don't have an update.

I don't know what's going wrong for you; a bunch of people are using XLA
internally without issue.  It's pretty surprising.

**Comment by joeyearsley** [comment]
Ok, thank you for the update!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by joeyearsley** [comment]
@jlebar I notice that from TF 1.12 that it will be built with XLA by default.

Is this really a good idea with a bug such as this is still unsolved?

**Comment by d0k** [comment]
> @jlebar I notice that from TF 1.12 that it will be built with XLA by default.
> 
> Is this really a good idea with a bug such as this is still unsolved?

It *builds* with XLA, it doesn't *run* with XLA by default. I'm quite unhappy with this bug still being around, but failed to ever reproduce the crash you're seeing.

**Comment by jlebar** [comment]
Xla will be built, but it will not be enabled in models unless you ask for
that.

**Comment by joeyearsley** [comment]
Fair points. 
@d0k do you have any hypothesis of what the bug could be? 

I’ve ignored the crash for now as I’m putting it down to some weird clash in the environment.

**Comment by d0k** [comment]
> @d0k do you have any hypothesis of what the bug could be?

The status you're seeing is CUDNN_STATUS_MAPPING_ERROR, which hints at an internal error in cudnn, so the bug is most likely there. But it's hard to say, there are a lot of moving parts.

**Comment by joeyearsley** [comment]
@tfboyd I noticed from the mailing list that you are experimenting a lot with XLA atm. Could you please try running the mnist-XLA repo I’ve posted above, do you see the same problems as myself?

**Comment by tfboyd** [comment]
I think I can make time to do that since I have a setup. Added myself to the assignee list for more visual nagging.  :-)  I run the ResNet example with tf.cnn.benchmarks a lot but I am using CUDA 10 because of the ptxas 9.0 vs 9.2 problem.  If you want to use CUDA 10 I and others have put up some binaries.  We will move TF nightly to CUDA 10 in mid/late November and then release it in January.  I wanted to go earlier but the timing did not work out.  

https://github.com/tensorflow/tensorflow/issues/22706

**Comment by tfboyd** [comment]
@joeyearsley  The docker image does not have the correct XLA due to some difficulties.  I need to see if I can pull it off with just apt-get, I cannot just copy the files over due to process and previous legal issues that may be getting resolved.

**Comment by joeyearsley** [comment]
Ah ok, thanks for the help and updates!

**Comment by tensorflowbutler** [comment]
Nagging Assignees @shivaniag, @tfboyd: It has been 33 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
